{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ecce3a7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   chrom     start       end         name  score strand  thickStart  thickEnd  \\\n",
      "0  chr22  10738828  10739148  PB.118859.1      0      -    10739148  10739148   \n",
      "1  chr22  10941696  10961547   PB.83093.1      0      -    10961547  10961547   \n",
      "2  chr22  11298750  11300438   PB.52878.1      0      +    11300438  11300438   \n",
      "3  chr22  11868469  11974188   PB.96068.1      0      -    11974188  11974188   \n",
      "4  chr22  11908367  11908901  PB.111548.1      0      -    11908901  11908901   \n",
      "\n",
      "       itemRgb  blockCount                    blockSizes  \\\n",
      "0  233,150,122           1                          320,   \n",
      "1   238,106,80           8  84,87,115,58,126,70,183,265,   \n",
      "2  233,150,122           1                         1688,   \n",
      "3  102,194,164           3                   477,59,258,   \n",
      "4   65,182,196           1                          534,   \n",
      "\n",
      "                                blockStarts            annotation novelty  \n",
      "0                                        0,            intergenic   novel  \n",
      "1  0,3270,5607,7515,8352,17370,18635,19586,  novel_not_in_catalog   novel  \n",
      "2                                        0,            intergenic   novel  \n",
      "3                          0,103470,105461,             antisense   novel  \n",
      "4                                        0,          genic_intron   novel  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define the column names for the BED file\n",
    "column_names = [\n",
    "    \"chrom\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"thickStart\", \"thickEnd\",\n",
    "    \"itemRgb\", \"blockCount\", \"blockSizes\", \"blockStarts\", \"annotation\", \"novelty\"\n",
    "]\n",
    "\n",
    "# Read the BED file into a DataFrame\n",
    "file_path = \"./data/UHR_chr22_corrected.colored.bed\"\n",
    "df_bed = pd.read_csv(file_path, sep=\"\\t\", header=None, names=column_names)\n",
    "\n",
    "# Display the first few rows of the DataFrame\n",
    "print(df_bed.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c3e55f6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     isoform  chrom strand  length  exons      structural_category  \\\n",
      "0  PB.3796.1  chr22      +    5749      3     novel_not_in_catalog   \n",
      "1  PB.3796.2  chr22      +    5327      8     novel_not_in_catalog   \n",
      "2  PB.3797.1  chr22      +    6070      6                   fusion   \n",
      "3  PB.3798.1  chr22      -    6333      3  incomplete-splice_match   \n",
      "4  PB.3799.1  chr22      -    6053     21  incomplete-splice_match   \n",
      "\n",
      "                        associated_gene associated_transcript  ref_length  \\\n",
      "0                    ENSG00000206195.11                 novel      4136.0   \n",
      "1                    ENSG00000206195.11                 novel      2398.0   \n",
      "2  ENSG00000100181.22_ENSG00000283633.1                 novel      1019.0   \n",
      "3                    ENSG00000093100.13     ENST00000476405.1      4328.0   \n",
      "4                     ENSG00000243156.9     ENST00000441493.7      9447.0   \n",
      "\n",
      "   ref_exons  ...  seq_A_downstream_TTS  dist_to_CAGE_peak  within_CAGE_peak  \\\n",
      "0        4.0  ...  AAAAAAAAAAAAAAAAGAAC                NaN             False   \n",
      "1        8.0  ...  AAAACAAACAAACGAAAATA                NaN             False   \n",
      "2        4.0  ...  AAAAAAATAAAAAATTAAAA             -345.0             False   \n",
      "3        5.0  ...  AAATAAATAGTGTGTTTCTT                NaN             False   \n",
      "4       32.0  ...  AAAAAAAAAAAAAAAAAAGG              -22.0              True   \n",
      "\n",
      "   dist_to_polyA_site within_polyA_site  polyA_motif polyA_dist  \\\n",
      "0                 NaN               NaN       AATAAA      -14.0   \n",
      "1                 NaN               NaN          NaN        NaN   \n",
      "2                 NaN               NaN          NaN        NaN   \n",
      "3                 NaN               NaN       AATAAA       -9.0   \n",
      "4                 NaN               NaN          NaN        NaN   \n",
      "\n",
      "   polyA_motif_found                                            ORF_seq  \\\n",
      "0               True  MSRAGSHPKPAIPGRGGEKLALLLAMLGGRCCPAGGPESHTPRKPW...   \n",
      "1              False  MSRAGSHPKPAIPGRGGEKLALLLAMLGGRCCPAGGPESHTPRKPW...   \n",
      "2              False  MSVWCQDAPCGLCPGRSALAQCSFSVSVTCCLQSSSLAYLGRPLAH...   \n",
      "3               True  MGKKDDPKLMQEWFKLVQEKNAMVRYESELMIFARELELEDRQSRL...   \n",
      "4              False  MEERKHETMNPAHVLFDRFVQATTCKGTLKAFQELCDHLELKPKDY...   \n",
      "\n",
      "     ratio_TSS  \n",
      "0  2101.000000  \n",
      "1     9.955224  \n",
      "2          NaN  \n",
      "3     0.800399  \n",
      "4          NaN  \n",
      "\n",
      "[5 rows x 48 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Define the file path\n",
    "file_path = './data/UHR_chr22_classification.txt'\n",
    "\n",
    "# Read the file into a pandas DataFrame\n",
    "df_classification = pd.read_csv(file_path, sep='\\t')\n",
    "\n",
    "# Display the first few rows of the dataset\n",
    "print(df_classification.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bb04ff60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  chrom_x     start       end         name  score strand_x  thickStart  \\\n",
      "0   chr22  10738828  10739148  PB.118859.1      0        -    10739148   \n",
      "1   chr22  10941696  10961547   PB.83093.1      0        -    10961547   \n",
      "2   chr22  11298750  11300438   PB.52878.1      0        +    11300438   \n",
      "3   chr22  11868469  11974188   PB.96068.1      0        -    11974188   \n",
      "4   chr22  11908367  11908901  PB.111548.1      0        -    11908901   \n",
      "\n",
      "   thickEnd      itemRgb  blockCount  ...  seq_A_downstream_TTS  \\\n",
      "0  10739148  233,150,122           1  ...  AAAAAAAAAAAAAATAAAAG   \n",
      "1  10961547   238,106,80           8  ...  AATTACGTGTGTGTATTCTT   \n",
      "2  11300438  233,150,122           1  ...  TCAAAAAAAAAAAAAAAAGA   \n",
      "3  11974188  102,194,164           3  ...  AAAAAAAAAAAAAAAGCAGC   \n",
      "4  11908901   65,182,196           1  ...  CCAAAAAAAAAAAAAAAAAA   \n",
      "\n",
      "  dist_to_CAGE_peak within_CAGE_peak dist_to_polyA_site within_polyA_site  \\\n",
      "0               NaN            False                NaN               NaN   \n",
      "1             -39.0            False                NaN               NaN   \n",
      "2               NaN            False                NaN               NaN   \n",
      "3               NaN            False                NaN               NaN   \n",
      "4               NaN            False                NaN               NaN   \n",
      "\n",
      "  polyA_motif polyA_dist  polyA_motif_found  \\\n",
      "0         NaN        NaN              False   \n",
      "1      TATAAA      -14.0               True   \n",
      "2         NaN        NaN              False   \n",
      "3      AGTAAA      -29.0               True   \n",
      "4         NaN        NaN              False   \n",
      "\n",
      "                                             ORF_seq ratio_TSS  \n",
      "0                                                NaN       NaN  \n",
      "1  MDEGTYIHALNNGLFTLGAPHKEVDEGPSPPEQFTAVKLSDSRITL...       NaN  \n",
      "2                                                NaN       NaN  \n",
      "3  MSLRVGARAKRNPWASGDPGGPDQCPLVVGADAWAHCGRAGPEVQV...       NaN  \n",
      "4                                                NaN       NaN  \n",
      "\n",
      "[5 rows x 62 columns]\n"
     ]
    }
   ],
   "source": [
    "# Merge the two dataframes on the isoform and name columns\n",
    "merged_df = pd.merge(df_bed, df_classification, left_on='name', right_on='isoform')\n",
    "\n",
    "# Display the first few rows of the merged dataframe\n",
    "print(merged_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0f84a692",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       isoform chrom_x     start       end  score strand_x  thickStart  \\\n",
      "0  PB.118859.1   chr22  10738828  10739148      0        -    10739148   \n",
      "1   PB.83093.1   chr22  10941696  10961547      0        -    10961547   \n",
      "2   PB.52878.1   chr22  11298750  11300438      0        +    11300438   \n",
      "3   PB.96068.1   chr22  11868469  11974188      0        -    11974188   \n",
      "4  PB.111548.1   chr22  11908367  11908901      0        -    11908901   \n",
      "\n",
      "   thickEnd      itemRgb  blockCount                    blockSizes  \\\n",
      "0  10739148  233,150,122           1                          320,   \n",
      "1  10961547   238,106,80           8  84,87,115,58,126,70,183,265,   \n",
      "2  11300438  233,150,122           1                         1688,   \n",
      "3  11974188  102,194,164           3                   477,59,258,   \n",
      "4  11908901   65,182,196           1                          534,   \n",
      "\n",
      "                                blockStarts            annotation novelty  \\\n",
      "0                                        0,            intergenic   novel   \n",
      "1  0,3270,5607,7515,8352,17370,18635,19586,  novel_not_in_catalog   novel   \n",
      "2                                        0,            intergenic   novel   \n",
      "3                          0,103470,105461,             antisense   novel   \n",
      "4                                        0,          genic_intron   novel   \n",
      "\n",
      "  chrom_y strand_y  length  exons   structural_category  \\\n",
      "0   chr22        -     320      1            intergenic   \n",
      "1   chr22        -     988      8  novel_not_in_catalog   \n",
      "2   chr22        +    1688      1            intergenic   \n",
      "3   chr22        -     794      3             antisense   \n",
      "4   chr22        -     534      1          genic_intron   \n",
      "\n",
      "                  associated_gene associated_transcript  ref_length  \\\n",
      "0                   novelGene_124                 novel         NaN   \n",
      "1               ENSG00000283047.1                 novel       749.0   \n",
      "2                    novelGene_53                 novel         NaN   \n",
      "3  novelGene_ENSG00000286406.1_AS                 novel         NaN   \n",
      "4                   novelGene_111                 novel         NaN   \n",
      "\n",
      "   ref_exons  diff_to_TSS  diff_to_TTS  diff_to_gene_TSS  diff_to_gene_TTS  \\\n",
      "0        NaN          NaN          NaN               NaN               NaN   \n",
      "1        9.0          NaN          NaN             209.0           -2309.0   \n",
      "2        NaN          NaN          NaN               NaN               NaN   \n",
      "3        NaN          NaN          NaN               NaN               NaN   \n",
      "4        NaN          NaN          NaN               NaN               NaN   \n",
      "\n",
      "                     subcategory  RTS_stage all_canonical  min_sample_cov  \\\n",
      "0                      mono-exon      False           NaN             NaN   \n",
      "1  at_least_one_novel_splicesite      False     canonical             0.0   \n",
      "2                      mono-exon      False           NaN             NaN   \n",
      "3                     multi-exon      False     canonical             0.0   \n",
      "4                      mono-exon      False           NaN             NaN   \n",
      "\n",
      "   min_cov min_cov_pos  sd_cov  FL  n_indels  n_indels_junc   bite   iso_exp  \\\n",
      "0      NaN         NaN     NaN   2       NaN            NaN    NaN  80.02650   \n",
      "1      0.0  junction_1     0.0   2       NaN            NaN  False   0.00000   \n",
      "2      NaN         NaN     NaN   2       NaN            NaN    NaN   9.26765   \n",
      "3      0.0  junction_1     0.0   2       NaN            NaN   True   0.00000   \n",
      "4      NaN         NaN     NaN   3       NaN            NaN    NaN  22.61715   \n",
      "\n",
      "   gene_exp  ratio_exp FSM_class      coding  ORF_length  CDS_length  \\\n",
      "0  80.02650        1.0         A  non_coding         NaN         NaN   \n",
      "1   0.00000        NaN         A      coding       138.0       414.0   \n",
      "2   9.26765        1.0         A  non_coding         NaN         NaN   \n",
      "3   0.00000        NaN         A      coding       111.0       336.0   \n",
      "4  22.61715        1.0         A  non_coding         NaN         NaN   \n",
      "\n",
      "   CDS_start  CDS_end  CDS_genomic_start  CDS_genomic_end predicted_NMD  \\\n",
      "0        NaN      NaN                NaN              NaN           NaN   \n",
      "1      575.0    988.0         10950118.0       10941697.0         False   \n",
      "2        NaN      NaN                NaN              NaN           NaN   \n",
      "3       40.0    375.0         11974149.0       11868889.0         False   \n",
      "4        NaN      NaN                NaN              NaN           NaN   \n",
      "\n",
      "   perc_A_downstream_TTS  seq_A_downstream_TTS  dist_to_CAGE_peak  \\\n",
      "0                   90.0  AAAAAAAAAAAAAATAAAAG                NaN   \n",
      "1                   20.0  AATTACGTGTGTGTATTCTT              -39.0   \n",
      "2                   85.0  TCAAAAAAAAAAAAAAAAGA                NaN   \n",
      "3                   80.0  AAAAAAAAAAAAAAAGCAGC                NaN   \n",
      "4                   90.0  CCAAAAAAAAAAAAAAAAAA                NaN   \n",
      "\n",
      "   within_CAGE_peak  dist_to_polyA_site  within_polyA_site polyA_motif  \\\n",
      "0             False                 NaN                NaN         NaN   \n",
      "1             False                 NaN                NaN      TATAAA   \n",
      "2             False                 NaN                NaN         NaN   \n",
      "3             False                 NaN                NaN      AGTAAA   \n",
      "4             False                 NaN                NaN         NaN   \n",
      "\n",
      "   polyA_dist  polyA_motif_found  \\\n",
      "0         NaN              False   \n",
      "1       -14.0               True   \n",
      "2         NaN              False   \n",
      "3       -29.0               True   \n",
      "4         NaN              False   \n",
      "\n",
      "                                             ORF_seq  ratio_TSS  \n",
      "0                                                NaN        NaN  \n",
      "1  MDEGTYIHALNNGLFTLGAPHKEVDEGPSPPEQFTAVKLSDSRITL...        NaN  \n",
      "2                                                NaN        NaN  \n",
      "3  MSLRVGARAKRNPWASGDPGGPDQCPLVVGADAWAHCGRAGPEVQV...        NaN  \n",
      "4                                                NaN        NaN  \n"
     ]
    }
   ],
   "source": [
    "# Drop the 'name' column\n",
    "merged_df = merged_df.drop(columns=['name'])\n",
    "\n",
    "# Reorder columns to make 'isoform' the first column\n",
    "columns = ['isoform'] + [col for col in merged_df.columns if col != 'isoform']\n",
    "merged_df = merged_df[columns]\n",
    "\n",
    "# Display the first few rows of the updated dataframe\n",
    "print(merged_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7f59e7d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7374e64a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics_py_3_9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
